report/5.Solution/StructureCodebase.tex

```diff
@@ -0,0 +1,121 @@
+\section{Structure of Codebase}
+The codebase is split into two main parts, one part that is dedicated to object detection and the other part is for movement and interaction with the motors.
+\todo{figure out whether the python part also does the calibration or not}
+The object detection part, which is run on a Raspberry Pi, is written in Python 3{.}7, and the movement processing is handled by the NXT and is written i C, version C90 with GNU-extensions.
+
+
+\subsection{Object Detection}
+The codebase is seperated into modules with the intent of creating a system where each module can be exchanged based on different contexts and further development.
+For the project, the following modules are used:
+\begin{itemize}
+	\item Object Localization Algorithm
+	\item Output / Communication
+	\item Input / Webcam
+\end{itemize}
+
+These are all utilized by a controller class, called \texttt{FlatController}, which ties the solution together, and has the signature shown in Snippet~\ref{lst:FlatControllerInit}.
+\begin{lstlisting}[language=Python,label={lst:FlatControllerInit},caption={Initialization method of the \texttt{FlatController} class}]
+def __init__(self,
+		algorithm: Callable[[np.ndarray], Vector],
+		output_device: OutputDevice,
+		input_device: VideoController,
+		calibration_algorithm: Union[Callable[[], None], None] = None,
+	) -> None:
+	"""
+	Initializes the controller
+	
+	:param algorithm: The algorithm to use for image processing
+	:param output_device: The device to send data to
+	:param input_device: What type the capturing device should be.
+	:param calibration_algorithm: a function that takes calibration packages,
+	and handles them in some unknown way (either logs them or sends them back)
+"""
+
+\end{lstlisting}
+
+The order of the parameters does not correspond to the order of relevance, but rather is a requirement as two of the parameters have default arguments.
+This method takes, as the signature shows, a function, \texttt{algorithm} with an array as input and a vector as output.
```

Comment:
den her sætning kunne jeg godt tænke mig omformuleret, mest fordi at jeg ikke er sikker på hvad den egentlig siger.

Måske kan den og de næste to slåes sammen til én sætning der bare fortæller hvad input og output er